### PR TITLE
bpo-35588: Speed up mod, divmod and floordiv operations for Fraction type

### PR DESCRIPTION
**Pull request by scoder**
Implement mod and divmod operations for Fraction type by spelling out the numerator/denominator calculation, instead of instantiating and normalising Fractions along the way. This speeds up '%' and divmod() by 2-3x.

https://bugs.python.org/issue35588

### CODE REVIEW

File: Lib/fractions.py [hunk 1]
```diff
@@ -35,6 +35,17 @@ def _gcd(a, b):
         a, b = b, a%b
     return a
 
+
+def _flat_divmod(na, da, nb, db):
+    """(a // b, a % b)"""
+    # div = a // b
```
Comment (eamanu) [inline Lib/fractions.py]: why this comments?

File: Lib/fractions.py [hunk 2]
```diff
@@ -35,6 +35,17 @@ def _gcd(a, b):
...
+    n_div, d_div = na * db, da * nb
+    div = n_div // d_div
+    # mod = a - b * div == (na*db - da*nb * div) / (da*db)
```
Comment (eamanu) [inline Lib/fractions.py]: why this comments?

File: Lib/fractions.py [hunk 3]
```diff
@@ -35,6 +35,17 @@ def _gcd(a, b):
...
+    n_mod, d_mod = n_div - d_div * div, da*db
```
Comment (eamanu) [inline Lib/fractions.py]:
```suggestion
    n_mod, d_mod = n_div - d_div * div, da * db
```

File: Lib/fractions.py (hunk 2 above)
Comment (scoder) [reply Lib/fractions.py]: The comments are there to explain what the code does mathematically. I don't think it's obvious from the operations otherwise.

File: Lib/fractions.py (hunk 3 above)
Comment (scoder) [reply Lib/fractions.py]: Sure, fair enough.

File: Lib/fractions.py (hunk 2 above)
Comment (eamanu) [reply Lib/fractions.py]: Ok great

File: Lib/fractions.py [hunk 4]
```diff
@@ -35,6 +35,17 @@ def _gcd(a, b):
         a, b = b, a%b
     return a
 
+
+def _flat_divmod(na, da, nb, db):
+    """(a // b, a % b)"""
+    # div = a // b
+    n_div, d_div = na * db, da * nb
+    div = n_div // d_div
```
Comment (serhiy-storchaka) [inline Lib/fractions.py]: Would not be better to use `div, n_mod = divmod(n_div,  d_div)`?

File: Lib/fractions.py [hunk 5]
```diff
@@ -429,14 +440,23 @@ def _div(a, b):
...
+            a.numerator, a.denominator, b.numerator, b.denominator)
+        return Fraction(div), Fraction(n_mod, d_mod)
+
+    __divmod__, __rdivmod__ = _operator_fallbacks(_divmod, divmod)
+
     def _mod(a, b):
         """a % b"""
-        div = a // b
-        return a - b * div
+        _, n_mod, d_mod = _flat_divmod(
```
Comment (serhiy-storchaka) [inline Lib/fractions.py]: Since `div` is ignored, this operation can be optimized further if inline `_flat_divmod()` and use `//` for calculating `n_mod`.

File: Lib/fractions.py [hunk 6]
```diff
@@ -429,14 +440,23 @@ def _div(a, b):
...
-        return math.floor(a / b)
+        return (a.numerator * b.denominator) // (a.denominator * b.numerator)
 
     __floordiv__, __rfloordiv__ = _operator_fallbacks(_floordiv, operator.floordiv)
 
+    def _divmod(a, b):
+        """(a // b, a % b)"""
+        div, n_mod, d_mod = _flat_divmod(
+            a.numerator, a.denominator, b.numerator, b.denominator)
+        return Fraction(div), Fraction(n_mod, d_mod)
```
Comment (serhiy-storchaka) [inline Lib/fractions.py]: Should not the first item be just integer `div`?

File: Lib/fractions.py (hunk 4 above)
Comment (scoder) [reply Lib/fractions.py]: Good call, I just tried it. It's slightly slower for small numbers and slightly faster for very large numbers. Difficult to say where to draw the line here.

File: Lib/fractions.py (hunk 5 above)
Comment (scoder) [reply Lib/fractions.py]: Not sure what you mean. I don't see how to save operations here.

File: Lib/fractions.py (hunk 6 above)
Comment (eamanu) [reply Lib/fractions.py]: hmm if we are in fractions modules, why not return a fraction? BTW, yes, the first item should be a `int`

Comment (scoder) [reply Lib/fractions.py]: I wouldn't mind returning an `int`, but didn't want to change the return type from what it was before. It used to be calculated by the built-in `divmod()` function, which probably used generic division, thus returning the input type again (`Fraction`).

Comment (serhiy-storchaka) [reply Lib/fractions.py]: Currently `divmod()` returns an integer for fractions. In general, `divmod(a, b)` returns a pair `(a // b, a % b)`.

File: Lib/fractions.py (hunk 5 above)
Comment (serhiy-storchaka) [reply Lib/fractions.py]: Sorry, I meant using `%`. Something like the following (untested):
```python
da = a.denominator
db = b.denominator
return Fraction((a.numerator * db) % (b.numerator * da), da * db)
```

File: Lib/fractions.py (hunk 6 above)
Comment (scoder) [reply Lib/fractions.py]: Ah, right. Looks like I didn't run the tests that I added with the original implementation. I'll change that back.

File: Lib/fractions.py (hunk 5 above)
Comment (scoder) [reply Lib/fractions.py]: Ah, right. Nice. Yes, that makes it another bit faster.

Comment (scoder) [reply Lib/fractions.py]: Turns out that inlining the helper function gives another visible speed-up.

File: Lib/fractions.py [hunk 7]
```diff
@@ -429,14 +430,22 @@ def _div(a, b):
...
+        return div, Fraction(n_mod, da * db)
+
+    __divmod__, __rdivmod__ = _operator_fallbacks(_divmod, divmod)
+
     def _mod(a, b):
         """a % b"""
-        div = a // b
-        return a - b * div
+        da, db = a.denominator, b.denominator
+        return Fraction((a.numerator * db) % (b.numerator * da), da * db)
```
Comment (eamanu) [inline Lib/fractions.py:447]: Why do you create the da and db? why don't use directly b.denominator and a.denominator? :)

Comment (serhiy-storchaka) [reply Lib/fractions.py:447]: For clarity and for performance. Using a local variable is faster than looking up an attribute.

File: Lib/fractions.py [hunk 8]
```diff
@@ -35,6 +35,7 @@ def _gcd(a, b):
         a, b = b, a%b
     return a
 
+
```
Comment (serhiy-storchaka) [inline Lib/fractions.py]: Unrelated change.

File: Lib/test/test_fractions.py [hunk 9]
```diff
@@ -117,6 +117,12 @@ def assertTypedEquals(self, expected, actual):
         self.assertEqual(type(expected), type(actual))
         self.assertEqual(expected, actual)
 
+    def assertTypedTupleEquals(self, expected, actual):
+        """Asserts that both the types and values in the tuples are the same."""
+        self.assertIsInstance(actual, tuple)
+        self.assertEqual(list(map(type, expected)), list(map(type, actual)))
```
Comment (serhiy-storchaka) [inline Lib/test/test_fractions.py]: This is not needed. Instead add
```python
for a, e in zip(actual, expected):
    self.assertTypedEquals(a, e)
```
after the next line.

File: Lib/test/test_fractions.py [hunk 10]
```diff
@@ -371,6 +380,40 @@ def testArithmetic(self):
         self.assertEqual(p.numerator, 4)
         self.assertEqual(p.denominator, 1)
 
+    def testLargeArithmetic(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_fractions.py:382]: Oh, sorry, I was unclear. I wanted to see results of benchmarks for large numbers, not adding unittests for them. I think `testLargeArithmetic` is not needed.

File: Lib/fractions.py (hunk 7 above)
Comment (eamanu) [reply Lib/fractions.py:447]: Great. So, this way, why don't create an a.numerator and b.numerator local variable?

File: Misc/NEWS.d/next/Library/2018-12-26-10-55-59.bpo-35588.PSR6Ez.rst [hunk 11]
```diff
@@ -0,0 +1,2 @@
+Mod, divmod and floordiv operations on :class:`fractions.Fraction` types are 2-4x faster.
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2018-12-26-10-55-59.bpo-35588.PSR6Ez.rst]:
```suggestion
 The floor division and modulo operations and the :func:`divmod` function on :class:`fractions.Fraction` types are 2--4x faster.
```

File: Lib/fractions.py (hunk 7 above)
Comment (asvetlov) [reply Lib/fractions.py:447]: I guess because `a.numerator` and `b.numerator` are used only once

**Review by asvetlov — APPROVED** [review]

**Review by eamanu — APPROVED** [review]
LGTM. @asvetlov I think is ready to merge.

File: Lib/fractions.py (hunk 7 above)
Comment (eamanu) [reply Lib/fractions.py:447]: good point.

File: Lib/test/test_fractions.py (hunk 9 above)
Comment (scoder) [reply Lib/test/test_fractions.py]: I would then also have to check that both tuples have the same length. Honestly, I prefer the current approach, or, in fact, the `assertTupleEqual()` and `assertListEqual()` methods, which I now switched to.

File: Lib/fractions.py (hunk 8 above)
Comment (scoder) [reply Lib/fractions.py]: Removed.

File: Lib/test/test_fractions.py (hunk 10 above)
Comment (scoder) [reply Lib/test/test_fractions.py:382]: Do you think it's bad to have these tests? I don't think exercising the implementation with larger values is useless.

File: Misc/NEWS.d/next/Library/2018-12-26-10-55-59.bpo-35588.PSR6Ez.rst [hunk 12]
```diff
@@ -0,0 +1,2 @@
+ The floor division and modulo operations and the :func:`divmod` function on :class:`fractions.Fraction` types are 2--4x faster.
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2018-12-26-10-55-59.bpo-35588.PSR6Ez.rst]: Leading spaces are redundant (GitHub glitch?).

**Review by serhiy-storchaka — APPROVED** [review]

File: Lib/test/test_fractions.py (hunk 10 above)
Comment (serhiy-storchaka) [reply Lib/test/test_fractions.py:382]: It is up to you.

File: Misc/NEWS.d/next/Library/2018-12-26-10-55-59.bpo-35588.PSR6Ez.rst (hunk 12 above)
Comment (scoder) [reply Misc/NEWS.d/next/Library/2018-12-26-10-55-59.bpo-35588.PSR6Ez.rst]: Removed.

**Comment by serhiy-storchaka** [comment]
I'll merge this PR after New Year if @rhettinger and @mdickinson do not have objections.
